Source/WebKit/Shared/glib/JavaScriptEvaluationResultGLib.cpp

```diff
@@ -0,0 +1,137 @@
+    m_map.add(identifier, toValue(variant));
+    return identifier;
+}
+
+auto JavaScriptEvaluationResult::toValue(GVariant* variant) -> Value
+{
+    if (g_variant_is_of_type(variant, G_VARIANT_TYPE("a{sv}"))) {
+        HashMap<JSObjectID, JSObjectID> map;
+        GVariantIter iter;
+        g_variant_iter_init(&iter, variant);
+        const char* key;
+        GVariant* value;
+        while (g_variant_iter_loop(&iter, "{&sv}", &key, &value)) {
+            if (!key || !value)
+                continue;
+            auto keyID = JSObjectID::generate();
+            m_map.add(keyID, String::fromUTF8(key));
+            auto valueID = JSObjectID::generate();
+            m_map.add(valueID, toValue(value));
+            map.add(keyID, valueID);
+        }
+        return { WTFMove(map) };
+    }
+
+    if (g_variant_is_of_type(variant, G_VARIANT_TYPE_UINT32))
+        return static_cast<double>(g_variant_get_uint32(variant));
+    if (g_variant_is_of_type(variant, G_VARIANT_TYPE_INT32))
+        return static_cast<double>(g_variant_get_int32(variant));
+    if (g_variant_is_of_type(variant, G_VARIANT_TYPE_UINT64))
+        return static_cast<double>(g_variant_get_uint64(variant));
+    if (g_variant_is_of_type(variant, G_VARIANT_TYPE_INT64))
+        return static_cast<double>(g_variant_get_int64(variant));
+    if (g_variant_is_of_type(variant, G_VARIANT_TYPE_INT16))
+        return static_cast<double>(g_variant_get_int16(variant));
+    if (g_variant_is_of_type(variant, G_VARIANT_TYPE_UINT16))
+        return static_cast<double>(g_variant_get_uint16(variant));
+    if (g_variant_is_of_type(variant, G_VARIANT_TYPE_DOUBLE))
+        return static_cast<double>(g_variant_get_double(variant));
+
+    if (g_variant_is_of_type(variant, G_VARIANT_TYPE_STRING))
+        return String::fromUTF8(g_variant_get_string(variant, nullptr));
+
+    return EmptyType::Null;
+}
+
+JavaScriptEvaluationResult::JavaScriptEvaluationResult(GVariant* variant)
+    : m_root(addObjectToMap(variant))
+{
+}
+
+static bool isSerializable(GVariant* variant)
+{
+    if (!variant)
+        return false;
+
+    if (g_variant_is_of_type(variant, G_VARIANT_TYPE_UINT32)
+        || g_variant_is_of_type(variant, G_VARIANT_TYPE_INT32)
+        || g_variant_is_of_type(variant, G_VARIANT_TYPE_UINT64)
+        || g_variant_is_of_type(variant, G_VARIANT_TYPE_INT64)
+        || g_variant_is_of_type(variant, G_VARIANT_TYPE_INT16)
+        || g_variant_is_of_type(variant, G_VARIANT_TYPE_UINT16)
+        || g_variant_is_of_type(variant, G_VARIANT_TYPE_DOUBLE)
+        || g_variant_is_of_type(variant, G_VARIANT_TYPE_STRING))
+        return true;
+
+    if (g_variant_is_of_type(variant, G_VARIANT_TYPE("a{sv}"))) {
+        GVariantIter iter;
+        g_variant_iter_init(&iter, variant);
+        const char* key;
+        GVariant* value;
+        while (g_variant_iter_loop(&iter, "{&sv}", &key, &value)) {
+            if (!key || !isSerializable(value))
+                return false;
```

Comment:
this leaks `value`, according to the `g_variant_iter_loop` documentation

Reply:
Thanks, created https://github.com/WebKit/WebKit/pull/50029